### PR TITLE
Replace oidc-client with oidc-client-ts

### LINKED ISSUE #177: The oidc-client has been archived
The library [oidc-client ](https://github.com/IdentityModel/oidc-client-js/wiki)which this library depends on, has recently been archived. So it isn't actively maintained anymore

Do you have any specific plans for the future regarding maintenance and continuation of this library?

### PR DESCRIPTION
**Pull request by mmizutani**
With big thanks to the maintainer of `oidc-client`, brockallen, this replaces the `oidc-client` library with its successor, [`oidc-client-ts`](https://github.com/authts/oidc-client-ts). [`oidc-client`](https://github.com/IdentityModel/oidc-client-js) is no longer maintained, and  it is not quite desirable to continue relying on abandoned libraries, especially for security sensitive authn/authz software.

The reasons that the migration the dependency from oidc-client to oidc-client-ts can be justified are as follow:

- oidc-client-ts is actively maintained:
  https://github.com/authts/oidc-client-ts/pulse
- The [APIs](https://authts.github.io/oidc-client-ts/) are mostly compatible with those of oidc-client. 
- They provide a detailed migration guide for the users of oidc-client:
  https://github.com/authts/oidc-client-ts/blob/main/docs/migration.md
- The maintainer of oidc-client recommends migration to oidc-client-ts:
  https://github.com/IdentityModel/oidc-client-js#:~:text=A%20successor%20project%20that%20is%20showing%20great%20progress%20in%20updating%20and%20modernizing%20is%20%22oidc%2Dclient%2Dts%22%20and%20can%20be%20found%20here.
  > A successor project that is showing great progress in updating and modernizing is "oidc-client-ts" and can be found [here](https://github.com/authts/oidc-client-ts).
- oidc-client and oidc-client-ts are both licensed under the Apache License 2.0. No license issues.

A downside is that oidc-client-ts does not support the OAuth 2.0 Implicit Flow.
oidc-client-ts suports the Implicit Flow and the Authorization Code Flow with PKCE, while oidc-client-ts only supports the Authorization Code Flow with PKCE and Refresh Token Grant.
Existing users of vuex-oidc might need to tweak some configurations of their identity providers to change the auth flow type.
However, OIDC service providers such as Microsoft and Auth0 nowadays recommend against the use of the Implicit Flow and encourage switching to the Authorization Code Flow with PKCE 
(or some other alternatives) even for SPAs. [Microsoft Authentication Library for JavaScript (MSAL.js) v2](https://github.com/AzureAD/microsoft-authentication-library-for-js), for example, has dropped support for the OAuth 2.0 Implicit Grant Flow.

In one of my Vue.js SPAs at work, I have experimentally switched from vuex-oidc with oidc-client to vuex-oidc with oidc-client-ts. The migration was quite smooth; I only needed to enable refresh tokens in AzureAD and slightly change the configurations for vuex-oidc as follows: 

```diff
import { vuexOidcCreateStoreModule } from 'vuex-oidc'
- import Oidc from 'oidc-client'
+ import { Logger, Log } from 'oidc-client-ts'

- Oidc.Log.logger = window.console
- Oidc.Log.level = process.env.NODE_ENV === 'development' ? Oidc.Log.DEBUG : Oidc.Log.WARN
+ Log.setLogger(window.console)
+ if (process.env.NODE_ENV === 'development') {
+   Log.setLevel(Log.DEBUG)
+ } else {
+   Log.setLevel(Log.WARN)
+ }

export const oidcSettings = {
  authority: `https://${process.env.oidcHost}/v2.0`,
  clientId: process.env.clientId,
  redirectUri: process.env.redirectUri,
- responseType: 'id_token token',
+ responseType: 'code',
- scope: 'openid profile',
+ scope: 'openid profile offline_access',
  automaticSilentRenew: true,
  validateSubOnSilentRenew: true,
  automaticSilentSignin: true,
  includeIdTokenInSilentRenew: true,
- revokeAccessTokenOnSignout: true
+ revokeTokensOnSignout: true
}
```

This has been working well in an Authorization Code Flow with PKCE and access token auto-refresh setup.

To help migration by the existing users of the vuex-oidc, I have also prepared some changes to the wiki of this repository:

https://github.com/mmizutani/vuex-oidc/wiki/Home/_compare/d9bb6cb0855e83464435bbf2042b7eac48f5b48c...d6da010b8af3a8ac2b1a2a8156f05972624f50e2

If the backward-incompatible changes introduced by this PR is not acceptable, we might consider adding a user-configurable parameter to select which of the OIDC client libraries to use to vuex-oidc.

Resolve #177

### CODE REVIEW

**Comment by mmizutani** [comment]
@perarnborg
I don't want to bother you, but could you please let us know what you think about this change to the core dependency of vuex-oidc?

**Comment by perarnborg** [comment]
Thank you for your time and effort put into this @mmizutani!

This does introduce some breaking changes. That could be possible in a major version bump though, so I am not completely against it. The added complexity of providing the option to choose dependency library (oidc-client / oidc-client-ts) does not appeal to me as a maintainer.

For me personaly this is a low priority, since I am part of a team building an alternative client lib to oidc-client-ts. The motivation of doing this is that there are some things inherited from oidc-client that I want to change for my projects. This library will also be publicly available, and will come with it's own wrappers for different frameworks like Vue. I am not planning on moving this project to that client, due to backward incompatibilities.

@mmizutani An alternative that I see is that we fork this into a different project (vuex-oidc-ts?). If you (or someone else in the community) would like to do that, feel free to grab the source code from this library if you want! Since my main focus is somewhat elsewhere for now, that could be an idea.

However, as I said I am also open to merging this and publish it as v4 as you suggest. The changes that where needed were indeed minimal. I am not completely ruling out the route of providing the option to choose dependency lib, it just does not appeal to me at first glance... I will need to think about this for a little while, please provide your feedback below! ❤️ 

For those interested, here is the other oauth/oidc client project that I am working on: https://github.com/Aventyret/lionel-oauth-client It is a work in progress.

**Comment by perarnborg** [comment]
@mediafreakch Good point about Google, this should be noted in the docs after v4 is published.

Did you (or anyone else here) test `authenticateOidcSilent`? It throws an error for me in the example application.

I am looking in to it, but I would appreciate any help given!

**Comment by perarnborg** [comment]
...okay, I realized there was another breaking change. `vuexOidcProcessSilentSignInCallback` now needs the oidcSettings as an argument. This is fine, I will add that to the docs as well.

I will merge these changes in the next couple of days and publish to npm as v4.

**Review by perarnborg — CHANGES_REQUESTED** [review]
@mmizutani Thank you for your contribution. This works as expected and I am ready to merge this.

Just a small comment on one of the changes, see line comment.

Since your branch has changes in the CHANGELOG.md file there are conflicts blocking the merge. Can you resolve these?

Otherwise (if you do not have the time at the moment) I can recreate your changes in a new branch.

File: src/services/oidc-helpers.js
Comment (mmizutani) [reply src/services/oidc-helpers.js]: Thanks. Yes, indeed. Addressed by https://github.com/perarnborg/vuex-oidc/pull/187/commits/4378aa4a505f4cee541da7b77802ff8899cd833d.

**Comment by mmizutani** [comment]
> Bottom line is that the Authorization Code flow with PKCE for Google requires you to send a clientSecret
> To make it work with Google, you would need an additional clientSecret which I did not know

> Did you (or anyone else here) test authenticateOidcSilent? It throws an error for me in the example application.

Sorry, I forgot to pass the `clientSecret` and `client_secret` from the `oidcConfig` to the `UserManager`. This is fixed in https://github.com/perarnborg/vuex-oidc/pull/187/commits/b7dfa86b85fb8533f9c2545bbc5998a262e3769a.

Here is a working example adapted from https://github.com/perarnborg/vuex-oidc-example that uses `oidc-client-ts` and this PR branch: https://github.com/mmizutani/vuex-oidc-example/tree/switch-to-oidc-client-ts
Since example app now requires not only the OAuth client ID but also the client secret, I have swapped the OAuth client credential with my own.

**Comment by perarnborg** [comment]
@mmizutani Thank you for moving so fast on this! Your changes look great, I plan to pubslish v4 on Monday after I have looked through everything one last time and after I have updated the docs.

**Review by perarnborg — APPROVED** [review]
Good job! 👏

**Comment by mmizutani** [comment]
Thank you for the discussion and review.

**Pull request merged**
